pkg/postage/batchstore/reserve.go

```diff
@@ -0,0 +1,241 @@
+}
+
+// unreserve is called when the batchstore decides not to reserve a batch on a PO
+// i.e.  chunk of  the batch in bins [0 upto PO] (closed  interval) are unreserved
+func (s *store) unreserve(batchID []byte, radius uint8) error {
+	return s.unreserveFunc(batchID, radius)
+}
+
+// purgeExpired is called when there is 'settlement'
+func (s *store) purgeExpired() error {
+	var toDelete [][]byte
+	err := s.store.Iterate(valueKeyPrefix, func(key, _ []byte) (bool, error) {
+		b, err := s.Get(valueKeyToID(key))
+		if err != nil {
+			return true, err
+		}
+		if b.Value.Cmp(s.rs.Core) >= 0 {
+			return true, nil
+		}
+		err = s.unreserve(b.ID, swarm.MaxPO)
+		if err != nil {
+			return true, err
+		}
+		// if batch has no value then delete it
+		if b.Value.Cmp(s.cs.Total) <= 0 {
+			toDelete = append(toDelete, b.ID)
+		} else {
+			b.Radius = swarm.MaxPO
+			if err := s.store.Put(batchKey(b.ID), b); err != nil {
+				return true, err
+			}
+		}
+
+		return false, nil
+	})
+	if err != nil {
+		return err
+	}
+	return s.delete(toDelete...)
+}
+
+// change captures the batch's change in value
+type change struct {
+	old, new int
+}
+
+func newChange(oldv, newv, lower, higher *big.Int) *change {
+	return &change{
+		old: cmp(oldv, lower, higher),
+		new: cmp(newv, lower, higher),
+	}
+}
+
+// cmp checks where value x of a batch falls wrt core and edge layer limit
+// it returns
+// - -1 if x is out of the reserve
+// - 0 if x is in the mid range where chunks are kept within depth
+// - 1 if x is in the top range where chunks are kept within depth - 1
+func cmp(x, lower, higher *big.Int) int {
+	if x.Cmp(lower) < 0 || x.Cmp(big.NewInt(0)) == 0 {
+		return -1
+	}
+	if x.Cmp(higher) < 0 {
+		return 0
+	}
+	return 1
+}
+
+// if value changed wrt range
+func (ch *change) changed() bool {
+	return ch.old != ch.new
+}
+
+// if the value is increased
+func (ch *change) increased() bool {
+	return ch.old < ch.new
+}
+
+// if value changed from -1 to 1 or 1 to -1
+func (ch *change) double() bool {
+	return ch.old+ch.new == 0
+}
+
+// if currently out of reserve, ie., new value falls in bottom range
+func (ch *change) out() bool {
+	return ch.new < 0
+}
+
+// updateReserveState manages what chunks of which batch are allocated to the reserve
+func (s *store) updateValueChange(b *postage.Batch, oldDepth uint8, oldValue *big.Int) error {
+	newValue := b.Value
+	newDepth := b.Depth
+
+	ch := newChange(oldValue, newValue, s.rs.Core, s.rs.Edge)
+	half := exp2(newDepth - s.rs.Depth - 1)
+	// if value drops out then available capacity is increased with the batch's size within neighbourhood
+	if ch.out() {
+		if ch.changed() {
+			s.rs.Capacity += half
+			if ch.double() {
+				s.rs.Capacity += half
+			}
+		}
+		return s.unreserve(b.ID, swarm.MaxPO)
+	}
+	// if value decreased and fell below the edge limit, then the freed capacity is added
+	if !ch.increased() {
+		s.rs.Capacity += half
```

Comment:
since you removed the `if !changed { return nil }` check earlier this can execute also if value remains the same.